layers: Add check for swapchain image acquire/sync

karl-lunarg opened the pull request:
Fixes #412

Issue validation error message at QueueSubmit time when swapchain
images referenced in a command buffer are not acquired or
associated with an acquire semaphore.

mikes-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2637,6 +2637,25 @@ static bool ValidateFenceForSubmit(layer_data *dev_data, FENCE_NODE *pFence) {
+                skip |=
+                    log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_COMMAND_BUFFER_EXT,
+                            HandleToUint64(pCB->commandBuffer), __LINE__, DRAWSTATE_SWAPCHAIN_IMAGE_NOT_ACQUIRED, "DS",
+                            "vkQueueSubmit: Swapchain image (0x%" PRIx64 ") has not been acquired or has not been synchronized.",
```
Log message could be more specific if we split it up for each error condition

karl-lunarg replied:
Yes, I realized the same a few minutes after I pushed the commit.  I think that is a good idea and will implement that.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2637,6 +2637,25 @@ static bool ValidateFenceForSubmit(layer_data *dev_data, FENCE_NODE *pFence) {
+        if (image_data && (image_data->binding.mem == MEMTRACKER_SWAP_CHAIN_IMAGE_KEY) && !image_data->shared_presentable) {
+            if (!image_data->acquired || (unsignaled_semaphores.find(image_data->semaphore) == unsignaled_semaphores.end())) {
+                skip |=
+                    log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT, VK_DEBUG_REPORT_OBJECT_TYPE_COMMAND_BUFFER_EXT,
```
No, I wanted to see if this solution was on the right track first.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2637,6 +2637,25 @@ static bool ValidateFenceForSubmit(layer_data *dev_data, FENCE_NODE *pFence) {
+    for (auto cb_image_data : pCB->imageLayoutMap) {
+        auto image_data = GetImageState(dev_data, cb_image_data.first.image);
+        if (image_data && (image_data->binding.mem == MEMTRACKER_SWAP_CHAIN_IMAGE_KEY) && !image_data->shared_presentable) {
+            if (!image_data->acquired || (unsignaled_semaphores.find(image_data->semaphore) == unsignaled_semaphores.end())) {
```
You're right, I'll need to look into this.

karl-lunarg commented on layers/core_validation.cpp:
```diff
@@ -2637,6 +2637,25 @@ static bool ValidateFenceForSubmit(layer_data *dev_data, FENCE_NODE *pFence) {
 }
 
+static bool validateImageAcquireAndSync(layer_data *dev_data, GLOBAL_CB_NODE *pCB,
+                                        unordered_set<VkSemaphore> unsignaled_semaphores) {
```
I saw the same sort of data being passed this way in another function, but that doesn't make it right.  I will fix, if needed.

karl-lunarg commented:
I'm now recording the fence used at acquire time in the image data so I can tell if a fence was used at acquire when checking the image at submit time.  If there is a fence, I skip the semaphore check, assuming that the app is using the fence for synchronization.  Does it make sense to actually try to see if the app checked the fence by calling WaitForFences or GetFenceStatus?

I also split the error message as suggested, fixed the unordered_set argument and changed the code in acquire to always update the semaphore and fence in the image data.

It will be some time before I can write a test, so reviewers may wish to wait on that.

mark-lunarg commented:
@mikes-lunarg, are you cool with this one now?

mikes-lunarg commented:
Updates look good.  Initially I was asking for a test case, but that might be unreasonable given the current lack of VkSurfaceKHR support in the test framework.

karl-lunarg commented:
@mikes-lunarg :  Any thoughts on my previous question:

> Does it make sense to actually try to see if the app checked the fence by calling WaitForFences or GetFenceStatus?

I'm sort of on the fence about it.  :-)

Basically, the code now assumes that if a fence was specified on the acquire, then the application made some sort of attempt to check it.

chrisforbes commented:
I'm uneasy about this. It appears to work for "some typical call sequences" but isn't rigorous.
